### PR TITLE
Add FeeRate addition and subtraction traits

### PR DESCRIPTION
**Pull request by yancyribbens**
I can't think of a reason to not support addition and subtraction with `fee_rate`.  If two fee rates are added, the resulting units `sat/wu` units are viable.

Without this, I have to do some shenanigans like:

```rust
let tmp = fee_rate.to_sat_per_kwu();
fee_rate = FeeRate::from_sat_per_kwu(tmp - 1);
```

### CODE REVIEW

**Comment by apoelstra** [comment]
Yeah, this seems reasonable. Concept ACK. Though we should implement the full matrix of `&self` `self` `&other` `other`.

**Comment by yancyribbens** [comment]
> Yeah, this seems reasonable. Concept ACK. Though we should implement the full matrix of &self self &other other.

Sure.  I'm looking at `Weight` for example and I see `AddAssign` which takes `&mut self`, although I don't see `other`.  In Amount there is `other` but that looks to be used in place of `rhs`.  Is `Weight` missing something also?

**Comment by apoelstra** [comment]
> Is `Weight` missing something also?

Probably yeah. I think everything (well, all newtypes around primitive `Copy` types) should implement all four.

**Comment by yancyribbens** [comment]
Aright I'll look at adding them.  I also didn't bother with `checked` variants since I'm just using in the test module and I don't care if it panics, although I guess adding the checked variants makes sense as well.  Marking as a draft for now.

**Comment by tcharding** [comment]
This is relate to https://github.com/rust-bitcoin/rust-bitcoin/issues/1639

**Comment by yancyribbens** [comment]
Yeah, this seems reasonable. Concept ACK. Though we should implement the full matrix of &self self &other other.

@apoelstra I've added `AddAssign`, `SubAssign`, `checked_add` and `checked_sub` to FeeRate.  Is there anything else I'm missing?

**Comment by tcharding** [comment]
> Though we should implement the full matrix of &self self &other other.

I believe what @apoelstra meant was that the following test should pass but it currently does not build.

```rust
    #[test]
    fn can_add() {
        let one = FeeRate(1);
        let two = FeeRate(2);
        let three = FeeRate(3);

        assert!(one + two == three);
        assert!(&one + &two == three);
        assert!(&one + two == three);
        assert!(one + &two == three);
    }
```

Note that for stdlib types this works
```rust
        let one = 1;
        let two = 2;
        let three = 3;
        ...
```

**Comment by apoelstra** [comment]
Yes, that's what I meant.

Though having `*Assign` traits is also good and something we should do.

**Comment by yancyribbens** [comment]
Alright, I see I think.  back to draft until I can add that as well.

**Comment by tcharding** [comment]
Looks like we'll need to shoosh the linter in the tests. The other two CI fails are fixed by #3395.

Since we have gone this fare I suppose we should support `f += &FeeRate(1)`  .

**Comment by yancyribbens** [comment]
> Looks like we'll need to shoosh the linter in the tests. The other two CI fails are fixed by https://github.com/rust-bitcoin/rust-bitcoin/pull/3395.

rebased off that branch for now

> Since we have gone this fare I suppose we should support f += &FeeRate(1) .

sure, I added this as well.

**Review by tcharding — APPROVED** [review]
ACK f60227b936fa5fbf636a026af487f8c637980e4d

**Comment by tcharding** [comment]
My ack is just for the last 6 patches, this will need rebase before we can merge.

**Comment by yancyribbens** [comment]
> My ack is just for the last 6 patches, this will need rebase before we can merge. Sorry to hold you up.

No problem.  Thanks.

**Comment by apoelstra** [comment]
I will wait until the rebased version to review.

**Comment by yancyribbens** [comment]
> I will wait until the rebased version to review.

Rebased.

**Comment by yancyribbens** [comment]
looks like the fix is not yet in master.  No worries, I'll try rebasing later when it passes.

**Review by apoelstra — APPROVED** [review]
ACK cb2146d5fac37ea7f68caa7cb92a0e44e1bb4c41 successfully ran local tests

**Review by tcharding — APPROVED** [review]
ACK cb2146d5fac37ea7f68caa7cb92a0e44e1bb4c41

**Pull request merged**
